* Name : Kale Aniket Devidas
* Recommendation System    

In [3]:
import pandas as pd
from sklearn.cross_validation import train_test_split
import numpy as np

In [18]:
cols = ['user_id' , 'age', 'sex', 'occupation', 'zip_code']
users = pd.read_csv("ml-100k/u.user",names=cols, sep="|")
users.head()

user_id  age sex  occupation zip_code
0        1   24   M  technician    85711
1        2   53   F       other    94043
2        3   23   M      writer    32067
3        4   24   M  technician    43537
4        5   33   F       other    15213

In [12]:
r_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings = pd.read_csv('ml-100k/u.data', sep='\t', names=r_cols,
 encoding='latin-1')
ratings.head()

user_id  movie_id  rating  unix_timestamp
0      196       242       3       881250949
1      186       302       3       891717742
2       22       377       1       878887116
3      244        51       2       880606923
4      166       346       1       886397596

In [17]:
i_cols = ['movie id', 'movie title' ,'release date','video release date', 'IMDb URL', 'unknown', 'Action', 'Adventure',
 'Animation', 'Children\'s', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy',
 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']
items = pd.read_csv('ml-100k/u.item', sep='|', names=i_cols, encoding='latin-1')
items.head()

movie id        movie title release date  video release date  \
0         1   Toy Story (1995)  01-Jan-1995                 NaN   
1         2   GoldenEye (1995)  01-Jan-1995                 NaN   
2         3  Four Rooms (1995)  01-Jan-1995                 NaN   
3         4  Get Shorty (1995)  01-Jan-1995                 NaN   
4         5     Copycat (1995)  01-Jan-1995                 NaN   

                                            IMDb URL  unknown  Action  \
0  http://us.imdb.com/M/title-exact?Toy%20Story%2...        0       0   
1  http://us.imdb.com/M/title-exact?GoldenEye%20(...        0       1   
2  http://us.imdb.com/M/title-exact?Four%20Rooms%...        0       0   
3  http://us.imdb.com/M/title-exact?Get%20Shorty%...        0       1   
4  http://us.imdb.com/M/title-exact?Copycat%20(1995)        0       0   

   Adventure  Animation  Children's   ...     Fantasy  Film-Noir  Horror  \
0          0          1           1   ...           0          0       0   
1          1          0           0   ...           0          0       0   
2          0          0           0   ...           0          0       0   
3          0          0           0   ...           0          0       0   
4          0          0           0   ...           0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 24 columns]

In [19]:
items.shape

(1682, 24)

# Lets Understand the Data and shape of data

In [20]:
ratings.shape

(100000, 4)

* lets divide rating data into train and test so we can build model on the training data.
* But luckily we had divided data so just we need to process the data.

In [25]:
r_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings_base = pd.read_csv('ml-100k/ua.base', sep='\t', names=r_cols, encoding='latin-1')
ratings_test = pd.read_csv('ml-100k/ua.test', sep='\t', names=r_cols, encoding='latin-1')
print(ratings_base.shape) 
print(ratings_test.shape)

(90570, 4)
(9430, 4)


# Now lets start with Graph lab where we can build the modul using the Graphlab

In [28]:
import graphlab
train = graphlab.SFrame(rating_base)
test = graphlab.SFrame(ratings_test)

* Lets start with simple Popularity Module

In [30]:
popularity_module = graphlab.popularity_recommender.create(train, user_id= "user_id", item_id="movie_id" , target="rating")

Recsys training: model = popularity

Warning: Ignoring columns unix_timestamp;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 90570 observations with 943 users and 1680 items.

Data prepared in: 0.163902s

90570 observations to process; with 1680 unique items.

In [34]:
# get recommedation for first five user
popularity_recom = popularity_module.recommend(users=range(1,5))
popularity_recom.print_rows(num_rows=25)

+---------+----------+-------+------+
| user_id | movie_id | score | rank |
+---------+----------+-------+------+
|    1    |   1536   |  5.0  |  1   |
|    1    |   1656   |  5.0  |  2   |
|    1    |   1500   |  5.0  |  3   |
|    1    |   1293   |  5.0  |  4   |
|    1    |   1599   |  5.0  |  5   |
|    1    |   1467   |  5.0  |  6   |
|    1    |   1201   |  5.0  |  7   |
|    1    |   1189   |  5.0  |  8   |
|    1    |   1122   |  5.0  |  9   |
|    1    |   814    |  5.0  |  10  |
|    2    |   1536   |  5.0  |  1   |
|    2    |   1656   |  5.0  |  2   |
|    2    |   1500   |  5.0  |  3   |
|    2    |   1293   |  5.0  |  4   |
|    2    |   1599   |  5.0  |  5   |
|    2    |   1467   |  5.0  |  6   |
|    2    |   1201   |  5.0  |  7   |
|    2    |   1189   |  5.0  |  8   |
|    2    |   1122   |  5.0  |  9   |
|    2    |   814    |  5.0  |  10  |
|    3    |   1536   |  5.0  |  1   |
|    3    |   1656   |  5.0  |  2   |
|    3    |   1500   |  5.0  |  3   |
|    3    | 

* Did u notice for all users the item id in same order last 814 and first 1536

In [40]:
rating_base.groupby(by = "movie_id")["rating"].mean().sort_values(ascending = False).head(20)

movie_id
1500    5.000000
1293    5.000000
1122    5.000000
1189    5.000000
1656    5.000000
1201    5.000000
1599    5.000000
814     5.000000
1467    5.000000
1536    5.000000
1449    4.714286
1642    4.500000
1463    4.500000
1594    4.500000
1398    4.500000
114     4.491525
408     4.480769
169     4.476636
318     4.475836
483     4.459821
Name: rating, dtype: float64

In [42]:
# collabaritive Filtering Module
# train our module
item_sim_model = graphlab.item_similarity_recommender.create(train, user_id='user_id', item_id='movie_id', target='rating', similarity_type='pearson')

Recsys training: model = item_similarity

Warning: Ignoring columns unix_timestamp;

To use these columns in scoring predictions, use a model that allows the use of additional features.

Preparing data set.

Data has 90570 observations with 943 users and 1680 items.

Data prepared in: 0.174609s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 5.614ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 25.827ms                            | 0.25             | 6               |

| 390.406ms                           | 100              | 1680            |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 1.41326s

In [50]:
item_sim_recomm = item_sim_model.recommend(users=range(1,6), k=5)
item_sim_recomm.print_rows(num_rows=25)

+---------+----------+-------+------+
| user_id | movie_id | score | rank |
+---------+----------+-------+------+
|    1    |   1599   |  5.0  |  1   |
|    1    |   1201   |  5.0  |  2   |
|    1    |   1189   |  5.0  |  3   |
|    1    |   1122   |  5.0  |  4   |
|    1    |   814    |  5.0  |  5   |
|    2    |   1599   |  5.0  |  1   |
|    2    |   1201   |  5.0  |  2   |
|    2    |   1189   |  5.0  |  3   |
|    2    |   1122   |  5.0  |  4   |
|    2    |   814    |  5.0  |  5   |
|    3    |   1599   |  5.0  |  1   |
|    3    |   1201   |  5.0  |  2   |
|    3    |   1189   |  5.0  |  3   |
|    3    |   1122   |  5.0  |  4   |
|    3    |   814    |  5.0  |  5   |
|    4    |   1599   |  5.0  |  1   |
|    4    |   1201   |  5.0  |  2   |
|    4    |   1189   |  5.0  |  3   |
|    4    |   1122   |  5.0  |  4   |
|    4    |   814    |  5.0  |  5   |
|    5    |   1599   |  5.0  |  1   |
|    5    |   1201   |  5.0  |  2   |
|    5    |   1189   |  5.0  |  3   |
|    5    | 

In [55]:
model_performance = graphlab.compare(test, [popularity_module, item_sim_model])
graphlab.show_comparison(model_performance,[popularity_module, item_sim_model])

PROGRESS: Evaluate model M0

Precision and recall summary statistics by cutoff
+--------+-------------------+-------------------+
| cutoff |   mean_precision  |    mean_recall    |
+--------+-------------------+-------------------+
|   1    |        0.0        |        0.0        |
|   2    |        0.0        |        0.0        |
|   3    |        0.0        |        0.0        |
|   4    |        0.0        |        0.0        |
|   5    |        0.0        |        0.0        |
|   6    |        0.0        |        0.0        |
|   7    | 0.000151492198152 | 0.000106044538706 |
|   8    | 0.000132555673383 | 0.000106044538706 |
|   9    | 0.000235654530458 | 0.000212089077413 |
|   10   | 0.000212089077413 | 0.000212089077413 |
+--------+-------------------+-------------------+
[10 rows x 3 columns]

PROGRESS: Evaluate model M1

Precision and recall summary statistics by cutoff
+--------+-------------------+-------------------+
| cutoff |   mean_precision  |    mean_recall    |
+--